b3sum/src/main.rs

```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
     }
 }
 
+fn check_file(input: impl Read, mmap_disabled: bool) -> Result<u64> {
+    let hasher = blake3::Hasher::new();
+    let mut num_failed = 0;
+
+    let reader = std::io::BufReader::new(input);
+    for (line_num, line) in reader.lines().enumerate() {
+        // Read the line and trim it
+        let line: String = line.with_context(|| format!("line {}", line_num + 1))?;
+        let line = line.trim();
+
+        // Skip the line if it's a comment or empty
+        if line.starts_with('#') {
+            continue;
+        }
+        if line.len() == 0 {
+            continue;
+        }
+
+        // Split line by whitespace
+        let mut line = line.split_whitespace();
+
+        // Get the hash value and parse the hex value
+        let hash: &str = line.next().expect("Unexpected missing value");
+        let hash = hex::decode(hash)
+            .with_context(|| format!("error decoding hex on line {}", line_num + 1))?;
+
+        // Get the filename
+        let filename: &str = line
+            .next()
+            .ok_or_else(|| Error::msg(format!("missing filename on line {}", line_num + 1)))?;
+        let os_filename: std::ffi::OsString = filename.into();
```

Comment:
This raises an interesting point. Should we try to handle really weird filenames and escaping? It looks like `md5sum * | md5sum --check` can handle some very nasty stuff, including filenames with newlines or invalid unicode.

Reply:
I'm a big fan of drawing the line at non-unicode filenames.     Non-unicode filenames need to be relegated to the dustbin of history  :)  

But if you really think we need to support non-unicode filenames, then I can work on adding support. 

I agree that we should add tests for all sorts of pathological (unicode) filename cases though.

Reply:
On the one hand, it might not be worth the effort. On the other hand, if we treat everything as raw bytes, maybe we get this "support" for free? (But then again, Windows paths aren't amenable to raw bytes, are they? Not making Window support an annoyance might be a good reason to stick with Unicode/UTF-8.)

Reply:
Also, deciding what is whitespace and a line-end when interpreting raw bytes is a PITA.    I think I would want to bring in `regex::bytes` as a dependency to handle this if we decide to support non-unicode.